control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -3530,3 +3544,233 @@ func healthCheckIdentityProvider(ctx context.Context, hcp *hyperv1.HostedControl
+		if err != nil || len(describeSGResult.SecurityGroups) == 0 {
+			logger.Error(err, "Failed to fetch security group", "sgID", sgID)
+			return "", fmt.Errorf("failed to fetch security group (id: %s), code: %s", sgID, awsErrorCode(err))
+		}
+
+		sg = describeSGResult.SecurityGroups[0]
+		logger.Info("Created security group", "id", sgID)
+	}
+	ingressPermissions := supportawsutil.DefaultWorkerSGIngressRules(vpcCIDR, sgID, awssdk.StringValue(sg.OwnerId))
+	_, err = ec2Client.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
+		GroupId:       awssdk.String(sgID),
+		IpPermissions: ingressPermissions,
+	})
+	if err != nil {
+		if awsErrorCode(err) != "InvalidPermission.Duplicate" {
+			logger.Error(err, "Failed to set security group ingress rules")
+			return "", fmt.Errorf("failed to set security group ingress rules, code: %s", awsErrorCode(err))
+		}
+		logger.Info("WARNING: got duplicate permissions error when setting security group ingress permissions", "sgID", sgID)
+	}
+	return sgID, nil
+}
+
+func (r *HostedControlPlaneReconciler) destroyAWSDefaultSecurityGroup(ctx context.Context, hcp *hyperv1.HostedControlPlane) error {
+	logger := ctrl.LoggerFrom(ctx)
+	if msg, isValid := hasValidCloudCredentials(hcp); !isValid {
+		logger.Info("Skipping default SecurityGroup cleanup", "reason", msg)
+		return nil
+	}
+	describeSGResult, err := r.ec2Client.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{Filters: awsSecurityGroupFilters(hcp.Spec.InfraID)})
+	if err != nil {
+		return fmt.Errorf("cannot list security groups: %w", err)
+	}
+	if len(describeSGResult.SecurityGroups) == 0 {
+		return nil
+	}
+	sg := describeSGResult.SecurityGroups[0]
+
+	if len(sg.IpPermissions) > 0 {
+		if _, err = r.ec2Client.RevokeSecurityGroupIngressWithContext(ctx, &ec2.RevokeSecurityGroupIngressInput{
+			GroupId:       sg.GroupId,
+			IpPermissions: sg.IpPermissions,
+		}); err != nil {
+			return fmt.Errorf("failed to revoke security group ingress permissions for %s: %w", awssdk.StringValue(sg.GroupId), err)
+		}
+	}
+
+	if len(sg.IpPermissionsEgress) > 0 {
+		if _, err = r.ec2Client.RevokeSecurityGroupEgressWithContext(ctx, &ec2.RevokeSecurityGroupEgressInput{
+			GroupId:       sg.GroupId,
+			IpPermissions: sg.IpPermissionsEgress,
+		}); err != nil {
+			return fmt.Errorf("failed to revoke security group egress permissions for %s: %w", awssdk.StringValue(sg.GroupId), err)
+		}
+	}
+
+	if _, err = r.ec2Client.DeleteSecurityGroupWithContext(ctx, &ec2.DeleteSecurityGroupInput{
+		GroupId: sg.GroupId,
+	}); err != nil {
+		return fmt.Errorf("failed to delete security group %s: %w", awssdk.StringValue(sg.GroupId), err)
+	}
+	return nil
+
+}
+
+func awsErrorCode(err error) string {
+	if awsErr, ok := err.(awserr.Error); ok {
+		return awsErr.Code()
+	}
+	return ""
+}
+
+func hasValidCloudCredentials(hcp *hyperv1.HostedControlPlane) (string, bool) {
+	if hcp.Spec.Platform.Type != hyperv1.AWSPlatform {
+		return "", true
```

Comment:
@csrwng , why are you returning `isValid=true` in case the hostedcontrolplane is not of AWS type?

In case of other provider, it causes the hosted control plane controller to access AWS API, and therefore crash here:
https://github.com/openshift/hypershift/pull/2146/files#diff-3d4ffda0834e09a59a4d510523932e20b9234bcb9816e40258a0fb32a87a7006R3716-R3720

Reply:
https://github.com/openshift/hypershift/pull/2167